tests/unittests/test_upgrade.py

```diff
@@ -19,8 +17,15 @@
 import pytest
 
 from cloudinit.sources import pkl_load
+from cloudinit.sources.DataSourceAzure import DataSourceAzure
+from cloudinit.sources.DataSourceNoCloud import DataSourceNoCloud
 from tests.unittests.helpers import resourceLocation
 
+DSNAME_TO_CLASS = {
+    "Azure": DataSourceAzure,
```

Comment:
This is great, but leaves us open on all other sources. I'd like to see this test coverage catch to all others (which I recognize is probably not reasonable, since we don't actually test every datasource currently).

That can happen later though.

Reply:
@holmanb +1 on this sentiment and I think we should card that up as a goal.  Getting coverage of all  datasources will be hard as we don't have access/credentials to launch everywhere to get a sample pkl. I think it is minimally worthwhile for us to at least get coverage on the top 6 cloud datasources measured as far as commit frequency. To ensure that future commits don't introduce breakage.  Even this minimal test added here will show us that the NoCloud datasource is still exposed to a missing "seed_dir" initialization across pkl_load. We should shore up the gaps. On clouds where we have access, we can launch a focal image, downgrade to the original release cloud-init 20.1 on focal, cloud-init clean --logs --reboot and harvest the pkl to add to this unit test for that datasource.